Approved/implicitly-scoped-qubit-allocation.md

```diff
@@ -0,0 +1,342 @@
+---
+title: Implicitly-scoped qubit allocation
+description: Provides more convenient syntax for allocating qubits.
+author: Sarah Marshall
+date: September 28, 2020
+---
+
+# Proposal
+
+1. Add new `use` and `borrow` statements to allocate qubits that are valid until the end of the current block.
+2. Add new scope statement that does not require a preceding keyword, and is executed exactly once; equivalent to `if (true) { }`.
+3. Deprecate and remove existing `using` and `borrowing` block statements.
+
+# Justification
+
+Qubit allocation is currently awkward in Q# for two reasons:
+
+1. It requires creating a new block to define the lifetime of the qubit, even when the intended lifetime is the same as the current block.
+2. Allocating multiple qubits at the same time requires either tuple destructuring or nested blocks, both of which can be hard to read.
+
+As a quantum programming language, qubit allocation is at the core of Q#.
+Both the simple case of allocating a single qubit for the remainder of the current block, and more complicated scenarios involving multiple qubit variables that are both clean and borrowed and have varying lifetimes, should have clear, intuitive syntax that doesn't require boilerplate.
+
+# Description
+
+Qubits are a managed resource in Q#: when qubits are allocated, they are valid only within the new block declared at allocation, after which they are released and can no longer be used.
+This proposal makes qubit allocation in Q# always scoped to the current block.
+Instead of being released at the end of a newly-declared block, qubits are released at the end of the current block.
+It also extends block statement syntax to allow precise control of qubit lifetime when needed in more complicated scenarios.
+Finally, this proposal aims to be syntactically minimal and consistent with other binding statements like `let`, in contrast to alternatives that have more complicated syntax.
+
+## Current Status
+
+Currently, the only mechanism to allocate qubits in Q# is with `using` and `borrowing` block statements.
+These statements will be deprecated and removed, replaced with new `use` and `borrow` statements that do not require a block, as well as a new scope statement that can be used when a new block is desired.
```

Comment:
Since `use` and `borrow` can currently be variable names (though hopefully they aren't terribly common!), is this a breaking change? Would it make sense to have a opt-in to the new behavior (i.e.: a special Python-like `open` statement, a C#- or Rust-like `#` statement, or similar)?

Reply:
Yes, it's a breaking change. I'm not sure if we want to go through a keyword opt-in while we are pre-1.0. We should probably decide how we want to handle new keywords in general.

Reply:
Hm, I am not sure if the change in keyword makes sense. What are the arguments for it?

Reply:
Generally, I think keywords should use simple word forms. We say `return x;`, not `returning x;`, `open Foo;`, not `opening Foo;`, etc. There are no other keywords in Q# that end in -ing besides `using` and `borrowing`, so we will be gaining consistency with all other keywords, not just `let`.

In terms of consistency with other languages like C#, it does not seem like there is any consensus among other programming languages. While C# has `using`, F# has `use` instead. Java re-uses the `try` keyword, and Python has `with`. I think similarity with C# is not a strong enough argument to go with `using`.

Reply:
To flip the question: what reasons are there to name it `using` instead of `use`?

Reply:
@bettinaheim Based on discussions offline, it sounds like you are OK with `use` and `borrow` instead of `using` and `borrowing`. I think we should still decide how to handle the new keywords: should they immediately break code that has identifiers named `use` and `borrow`, should we have an opt-in mechanism, or should we deprecate identifiers named `use` and `borrow` for a period of time?

Reply:
To quickly just recapitulate: use and borrow could be more convenient to use and there is no strong case based on what is used in other languages. I am in generally worried about unnecessary changes, and I wouldn't want to consider "keyword tweaks" past 1.0. Given that we are still in preview, we can consider this one. @viveis for input as well.